keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -69,28 +69,14 @@
 
 ## Release Signoff Checklist
 
-<!--
-**ACTION REQUIRED:** In order to merge code into a release, there must be an
-issue in [kubernetes/enhancements] referencing this KEP and targeting a release
-milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
-of the targeted release**.
-
-For enhancements that make changes to code or processes/procedures in core
-Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
-Signoff checklist to be completed.
-
-Check these off as they are completed for the Release Team to track. These
-checklist items _must_ be updated for the enhancement to be released.
--->
-
 Items marked with (R) are required *prior to targeting to a milestone / release*.
 
 - [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
-- [ ] (R) KEP approvers have approved the KEP status as `implementable`
-- [ ] (R) Design details are appropriately documented
-- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [X] (R) KEP approvers have approved the KEP status as `implementable`
+- [X] (R) Design details are appropriately documented
+- [X] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
 - [X] (R) Graduation criteria is in place
-- [ ] (R) Production readiness review completed
+- [X] (R) Production readiness review completed
```

Comment:
Sorry - this is super invonvenient for both you and me, but given I can't comment on lines you didn't change I can only do that here:

<unrelated to PRR itself, though I wanted to understand some stuff>

L583: you're saying that for DaemonSet deployment the owner will be a pod. That seems really suspicious. What if I need to update the spec of DS? The pods will get recreated, the objects will get deleted and the new ones will get crated? That doesn't sound like the best option to me...
I think it should be the node (and potentially some controller that will be removing that OwnerRef if the node is no longer targetted by the driver...)


<PRR-related>
L758: if I disable the feature and never reenable again, then those object will keep staying in etcd. It's probably fine, because we expect to graduate it anyway, but probably worth pointing that out explicitly.

L781: Can you add explicit "Will be added before the transition to beta." (as you added later).

L824: While scalability is not required for Alpha, I really recommend you to think about this question: "*Will enabling / using this feature result in any new API calls?" now. 
If we implement it badly, it may not scale in large clusters (overload apiserver/etcd) which would block transition to GA. It's fine to do shortcuts/inefficient implementations for Alpha, but if you do that unconsciously it may require significant unneded changes earlier.
If you say "no", I can't block it, but I highly encourage you to spend a bit time on thinking about it and fill this in.

Reply:
> What if I need to update the spec of DS? The pods will get recreated, the objects will get deleted and the new ones will get crated?

Correct.

> That doesn't sound like the best option to me... I think it should be the node (and potentially some controller that will be removing that OwnerRef if the node is no longer targetted by the driver...)

Such a controller would have to be part of the driver deployment because Kubernetes itself has no knowledge about how the driver is generating these objects. This then brings us back to the "what if the driver has no chance to clean up" part: simply removing the driver and its operator may leave stale objects behind. I'm not sure how common this mode of operation will be (it was primarily meant for local ephemeral volumes, but now those are going to be handled via "generic ephemeral inline volumes"), so I'd rather keep it very simple and refine later if needed.

>L758:  if I disable the feature and never reenable again, then those object will keep staying in etcd. It's probably fine, because we expect to graduate it anyway, but probably worth pointing that out explicitly.

So they won't be garbage-collected because the type is disabled? Hmm, I already had such a warning in the paragraph: "However, the new objects will still remain in the etcd database." I've added a "..., they just won't be visible until the feature is enabled again."

> L781: Can you add explicit "Will be added before the transition to beta." (as you added later).

Done.

> L824: While scalability is not required for Alpha, I really recommend you to think about this question: "*Will enabling / using this feature result in any new API calls?" now.

This has been discussed extensively. I could try to summarize all of those discussions now, but the result would probably be incomplete and/or out-dated because we intend to do performance evaluations of the alpha implementation and then change the implementation before beta. So I'd rather skip this part for alpha.

Reply:
>> What if I need to update the spec of DS? The pods will get recreated, the objects will get deleted and the new ones will get crated?

> Correct.

I don't think this is a behavior we really want. I don't want to block the whole effort on that, but I would like the item to figuring out the behavior of disappearing capacity object on DS updates to really be solved for Beta.

So please add that to the Beta graduation criteria.

L610/L596: why those objects need to be periodically updated? It seems that we need to fetch data from drivers periodically, but if nothing has changed, the updates shouldn't even be sent (we should be watching those objects on provisioner side and if the value is the same in cache, just skip updates)


> This has been discussed extensively. I could try to summarize all of those discussions now, but the result would probably be incomplete and/or out-dated because we intend to do performance evaluations of the alpha implementation and then change the implementation before beta. So I'd rather skip this part for alpha.

I think that scalability of this feature is potentially questionable, so I would really like to avoid having something that won't work by-design. I'm fine with not having all the answers there, but at least answering the first question in this section (any new API calls) - should be answered at this point.

Reply:
> I don't think this is a behavior we really want. I don't want to block the whole effort on that, but I would like the item to figuring out the behavior of disappearing capacity object on DS updates to really be solved for Beta.
>
> So please add that to the Beta graduation criteria.

Okay, but let's make it an item about deciding what the desired behavior is and then implementing that.

> why those objects need to be periodically updated? It seems that we need to fetch data from drivers periodically, but if nothing has changed, the updates shouldn't even be sent (we should be watching those objects on provisioner side and if the value is the same in cache, just skip updates)

Correct, and that's how it is implemented - there is regular polling, but objects only get updated if changed. I'll update the text to make that clear.

> I think that scalability of this feature is potentially questionable

The problem is that scalability may be fine for some use-cases and not for others. I think adding such a feature is worthwhile, it just has to be well-understood and document what the feature is good for.

> I'm fine with not having all the answers there, but at least answering the first question in this section (any new API calls) - should be answered at this point.

Okay, I'll try to add something.

Reply:
I thought a bit more about this and came to the conclusion that the owner reference handling needs to become optional. We can kill several birds with one stone here: support drivers that are deployed outside of the cluster, reduce watch traffic by filtering which CSIStorageCapacity objects need to be sent to each external-provisioner instance, and avoid recreating all objects during a DaemonSet update. For that we need to introduce labels.

@msau42: can you look at the "storage capacity: optional owner reference + watch filter" commit (currently 33e9f28893bc2bd0ae90f8ee434ec90dad6887ad)? Does this make sense?

Reply:
Yes, that sounds good to me!